LanguageFeatures/Patterns/type_inference_A32_t01.dart

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion
+/// To type check a pattern p being matched against a value of type M:
+/// ...
+/// If p with required type T is in an irrefutable context:
+/// - It is a compile-time error if M is not assignable to T.
```

Comment:
This seems to be an alternative notion of 'assignable', cf. https://github.com/dart-lang/language/issues/2845#issuecomment-1434351830.

So we should understand the above spec line as "Perform applicable coercions on the matched value. It is a compile-time error if the static type of the coerced matched value is not assignable to `T`".

Consider the dynamic type check applied to an object which is obtained when the matched value type is `dynamic` and the required type is not a top type: The type check which is performed in this case is not strictly considered to be a coercion, because it is "included" in the notion of assignability ("`S` is assignable to `T` if `S <: T` or `S` is `dynamic`).